Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx

```diff
@@ -0,0 +1,382 @@
+    d->CancelButton->setEnabled(false);
+    d->SaveButton->setEnabled(false);
+    }
+  else if (updateText)
+    {
+    int position = d->TextEdit->textCursor().position();
+    std::string text = d->CurrentTextNode->GetText();
+    d->TextEdit->setText(text.c_str());
+    d->TextNodeModified = false;
+    QTextCursor cursor = d->TextEdit->textCursor();
+    position = std::min(position, d->TextEdit->toPlainText().length());
+    cursor.setPosition(position);
+    d->TextEdit->setTextCursor(cursor);
+    }
+  d->TextEdit->blockSignals(wasBlocking);
+
+  this->updateFinished();
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::setMRMLScene(vtkMRMLScene* scene)
+{
+  this->Superclass::setMRMLScene(scene);
+  this->updateRequested();
+}
+
+//------------------------------------------------------------------------------
+bool qMRMLTextWidget::isReadOnly()
+{
+  return this->ReadOnly;
+}
+
+//------------------------------------------------------------------------------
+bool qMRMLTextWidget::isEditing()
+{
+  Q_D(qMRMLTextWidget);
+  return d->isEditing();
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::setReadOnly(bool readOnly)
+{
+  Q_D(qMRMLTextWidget);
+  if (this->ReadOnly == readOnly)
+    {
+    return;
+    }
+
+  this->ReadOnly = readOnly;
+  this->readOnlyChanged(this->ReadOnly);
+  this->updateRequested();
+}
+
+//------------------------------------------------------------------------------
+bool qMRMLTextWidget::isAutoUpdate()
+{
+  Q_D(qMRMLTextWidget);
+  return this->AutoUpdate;
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::setAutoUpdate(bool autoUpdate)
+{
+  Q_D(qMRMLTextWidget);
+  if (this->AutoUpdate == autoUpdate)
+    {
+    return;
+    }
+
+  this->AutoUpdate = autoUpdate;
+  if (this->AutoUpdate)
+    {
+    d->setEditing(false);
+    this->onTextEditChanged();
+    }
+
+  this->autoUpdateChanged(this->AutoUpdate);
+  this->updateRequested();
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::onTextEditChanged()
+{
+  Q_D(qMRMLTextWidget);
+
+  d->TextEditModified = true;
+  if (!d->CurrentTextNode || !this->AutoUpdate)
+    {
+    return;
+    }
+
+  std::string text = d->TextEdit->toPlainText().toStdString();
+  d->CurrentTextNode->SetText(text.c_str(), VTK_ENCODING_UTF_8);
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::onEditButtonClicked()
+{
+  Q_D(qMRMLTextWidget);
+  d->TextEditModified = false;
+  d->TextEdit->setFocus();
+  d->setEditing(true);
+  this->updateRequested();
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::onCancelButtonClicked()
+{
+  Q_D(qMRMLTextWidget);
+
+  if (d->TextEditModified)
+    {
+    QMessageBox msgBox;
```

Comment:
Proper QMessageBox deleting is tricky. You may need to make it auto-close if you don't destroy explicitly. Search for other similar messageboxes and copy their implementation.

Reply:
This is how QMessageBox and ctkMessageBox are handled in qMRMLSceneViewMenu and qMRMLAnnotationTreeView.

Is there another style of implementation that would be better?